drivers/adc/adc_nxp_s32_adc_sar.c

```diff
@@ -0,0 +1,458 @@
+/*
+ * Copyright 2023 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/logging/log.h>
+#include <zephyr/drivers/adc.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/irq.h>
+#include <Adc_Sar_Ip_HwAccess.h>
+#include <Adc_Sar_Ip.h>
+#include <Adc_Sar_Ip_Irq.h>
+
+#define ADC_CONTEXT_USES_KERNEL_TIMER
+#include "adc_context.h"
+
+#define DT_DRV_COMPAT nxp_s32_adc_sar
+LOG_MODULE_REGISTER(adc_nxp_s32_adc_sar, CONFIG_ADC_LOG_LEVEL);
+
+/* ADC internal reference voltage in mV */
+#define ADC_NXP_S32_INTERNAL_VREF 3300
+/* Convert channel of group ADC to channel of physical ADC instance */
+#define ADC_NXP_S32_GROUPCHAN_2_PHYCHAN(group, channel)	\
+						(ADC_SAR_IP_HW_REG_SIZE * group + channel)
+
+struct adc_nxp_s32_config {
+	ADC_Type *base;
+	uint8_t group_channel;
+	uint8_t callback_select;
+	Adc_Sar_Ip_ConfigType *adc_cfg;
+	void (*irq_config_func)(const struct device *dev);
+	const struct pinctrl_dev_config *pin_cfg;
+};
+
+struct adc_nxp_s32_data {
+	const struct device *dev;
+	uint8_t instance;
+	struct adc_context ctx;
+	uint16_t *buffer;
+	uint16_t *buf_end;
+	uint16_t *repeat_buffer;
+	uint32_t mask_channels;
+	uint8_t num_channels;
+};
+
+static uint32_t adc_nxp_s32_get_instance(ADC_Type *base)
+{
+	uint32_t i;
+	ADC_Type *const base_ptrs[] = IP_ADC_BASE_PTRS;
+
+	for (i = 0; i < ARRAY_SIZE(base_ptrs); i++) {
+		if (base_ptrs[i] == base) {
+			break;
+		}
+	}
+
+	__ASSERT_NO_MSG(i < ARRAY_SIZE(base_ptrs));
+
+	return i;
+}
+
+static int adc_nxp_s32_init(const struct device *dev)
+{
+	const struct adc_nxp_s32_config *config = dev->config;
+	struct adc_nxp_s32_data *data = dev->data;
+	Adc_Sar_Ip_StatusType status;
+	/* This array shows max number of channels of each group */
+	uint8_t map_chan_group[ADC_SAR_IP_INSTANCE_COUNT][ADC_SAR_IP_NUM_GROUP_CHAN]
+							= FEATURE_ADC_MAX_CHN_COUNT;
+
+	data->instance = adc_nxp_s32_get_instance(config->base);
+	data->num_channels = map_chan_group[data->instance][config->group_channel];
```

Comment:
I also feel that "number of channels" sounds like it is probably a hardware property that could be described in DT

Reply:
yes, but it is defined in hardware header file. So I want  to take advantage of it, not add more DT prop for it.

Reply:
It isn't something that I would expect users to need to set but usually we have SOC level properties to account for hardware properties like this that might differ between SOC or HW version. These would not be set by a normal zephyr user but by someone enabling an SOC.

Reply:
In my implementation, ADC supports 3 group channels (precision, standard and external). So The number of channels that not only depend on the instance but also depend on the group channel https://github.com/zephyrproject-rtos/hal_nxp/pull/255/files#diff-7b513cb354fca8cd8087da2c98151d35b18a1e1916d88d2bde44773f0e73b1dcR123. So users still must set both "number of channels" and "group channel" DT prop, If use DT prop instead of getting hw header